### PR TITLE
Add record's __dict__ to structPayload

### PR DESCRIPTION
**Pull request by vmarkovtsev**
Not sure if it is not included intentionally... This is a light version of what is done at https://github.com/severb/graypy/blob/master/graypy/handler.py#L97

### CODE REVIEW

File: gcloud/logging/handlers/transports/background_thread.py [hunk 1]
```diff
@@ -127,7 +127,14 @@ def enqueue(self, record, message):
             self._entries_condition.acquire()
             if self.stopping:
                 return
+            record_dict = record.__dict__
+            for blacklisted in ("args", "msg", "message"):
```
Comment (dhermes) [inline gcloud/logging/handlers/transports/background_thread.py]: Use single quotes:

```python
('args', 'msg', 'message')
```

and below

```python
record_dict['exc_info']
```

Comment (dhermes) [inline gcloud/logging/handlers/transports/background_thread.py]: UPDATE: I just realized @waprin used double quotes when originally checking this code in.

Sigh.

**Comment by dhermes** [comment]
@waprin Can you review this?

**Comment by dhermes** [comment]
@vmarkovtsev See the [failing tests][1]. This likely will also need at least one new unit test. I'm happy to help if need be.

But we can hold off on tests until @waprin vets the usefulness of this code change.

[1]: https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/149487809

File: gcloud/logging/handlers/transports/background_thread.py (hunk 1 above)
Comment (waprin) [inline gcloud/logging/handlers/transports/background_thread.py]: Sorry, will def fix before merge into master.

Comment (dhermes) [inline gcloud/logging/handlers/transports/background_thread.py]: No worries. I blame the reviewer?

**Comment by vmarkovtsev** [comment]
Wow, cool! I didn't expect you would actually agree with this PR. I will fix the tests, code and add the new one for sure, thanks.

**Comment by dhermes** [comment]
@vmarkovtsev I didn't say I agree with it; I'm staying hands off until I hear from @waprin

**Comment by waprin** [comment]
@vmarkovtsev thanks for the PR, on the fence about this one but in general we want external contributions.

So a few issues I see:

1) This only addresses one transport. Currently there are two, this one and sync.py, and more will be added. They both inherit from a base class, and if we add this it should be generalized across the transports using the base class somehow.

2) What exactly is the justification for this? In the case of errors, we plan to add line number/file/stack trace to the Error Reporting API. Basically hook into [this](https://github.com/GoogleCloudPlatform/gcloud-python/blob/master/docs/error-reporting-usage.rst) when severity is an error. It's not clear it's something we want to add to every log message even if it's not an error. If you could provide specific examples, it might help.

3) This actually works differently than the example you linked to. It looks like greypy explicitly includes fields, then explicitly filters out most of the fields from the dict and only adds "extra" attributes `add_extra_fields`. I am not sure what type of extra fields they are expecting to find in the LogRecord (do you know)?. Either way, I would rather explicitly add thread id/process ID then add everything in the dict and filter out a blacklist.

**Comment by dhermes** [comment]
> Either way, I would rather explicitly add thread id/process ID then add everything in the dict and filter out a blacklist.

Ditto

**Comment by vmarkovtsev** [comment]
OK, this is my next attempt. If you say that it's not too radical I will fix the tests.

**Comment by vmarkovtsev** [comment]
BTW, extra_fields are as follows. You can add your custom LogRecord members using logging.Filter. E.g., I use the following code:
```python
import logging
import socket


class HostnameFilter(logging.Filter):
    hostname = socket.gethostname()

    def filter(self, record):
        record.hostname = self.hostname
        return True

handler = CloudLoggingHandler(gcloud.logging.Client(), extra_fields=('hostname',))
handler.addFilter(HostnameFilter())
```
Filters are not required to add new fields, they can delete, change, etc.

**Comment by vmarkovtsev** [comment]
I can make this "data" thing optional and turned off by default if you force "this adds overhead" thesis.

**Comment by waprin** [comment]
Hey sorry I missed these updates, will review tomorrow

File: gcloud/logging/handlers/handlers.py [hunk 2]
```diff
@@ -70,22 +77,54 @@ class CloudLoggingHandler(logging.StreamHandler, object):
...
+        :param message: The formatted log message
+
+        :return: dict with keyword arguments for
+                 :method:`gcloud.logging.Logger.log_struct()`.
+        """
+        return {
+            'info': {
+                'message': message,
+                'python_logger': record.name,
+                'data': {k: getattr(record, k, None) for k in itertools.chain(
```
Comment (waprin) [inline gcloud/logging/handlers/handlers.py]: for readability, probably better to refactor this above to a temp data variable that you just set here

**Comment by waprin** [comment]
Hey, thanks for the good example, it makes sense to me. We will want something like that added in the usage doc. 

The Cloud Console does support those kinds of filters, but it's still migrating from the v1 filter syntax to v2 filter syntax. In the meantime you can use the v2 syntax programmatically.

I'm not super concerned about overhead, just wanted to list things we should at least consider.

Overall I think you've justified why the change makes sense and I at least am happy to accept it (it's my feature branch but you will want signoff from one of the repo owners before merge). 

Do be aware that we need Python 2.7 and Python 3 support, so some things like itertools might require six. 

Thanks again, look forward to seeing Travis pass so we can merge.

File: gcloud/logging/handlers/handlers.py [hunk 3]
```diff
@@ -70,22 +77,54 @@ class CloudLoggingHandler(logging.StreamHandler, object):
 
     """
 
+    LOG_RECORD_DATA_WHITELIST = {
```
Comment (waprin) [inline gcloud/logging/handlers/handlers.py]: why is this a dict, not a list?

**Comment by vmarkovtsev** [comment]
Well, I had to merge with master because they dropped Python 2.6 and I have dict comprehension. But it screwed SLA (wtf) and I have no idea why Travis has problems.

**Comment by dhermes** [comment]
The CLA fail is likely because the merged in commits have authors other than you @vmarkovtsev. You should rebase or cherry-pick instead of merging and then force-push over this branch

**Comment by vmarkovtsev** [comment]
@dhermes Don't you say! And what's this? 17251f48836563ec58d32218c92f1aae2ab02665

**Comment by waprin** [comment]
@dhermes I think the issue is that @tseaver suggested using a feature branch but we never actually decided on a workflow that made sense unless you can give me write access to the feature branch, and I don't think Github has tha granular level of control. I could have developed the branch on my own repo but I wanted to make sure I was getting review at incremental points (and the fact that it was also part of this repo was helpful in getting some extra eyes on it like we're seeing in this PR getting made). 

The problem we keep seeing is that we need a repo owner to merge the master branch back into this one. Otherwise, if you rebase on top of master but try to merge into this branch you get all those extra commits. @vmarkovtsev  can't force push to the branch because again he doesn't have access.

Maybe the easiest thing to do is stick with personal repos for feature branch work and just ping you guys for reviews on it. In any case I would like to merge this feature branch into master soon anyway. There are lots of iterative improvements I would still like to make but none in my head will involve breaking changes. I wanted to add the fluentd handler but it's not even strictly necessary and will be a small non-breaking change. But if you'd really like me to wrap that up first it's no prob.

I asked @tseaver to merge my last PR for this branch but he had a few comments I still need to address and we should try to get this contribution merged as well. Then I'm wondering if there's any objection to just merging to master even if more work is queued up behind it.

**Comment by waprin** [comment]
@dhermes @tseaver another potential option: add me as collaborator to repo but setup master as a protected branch which you guys can just keep yourself on:

https://help.github.com/articles/configuring-protected-branches/

**Comment by waprin** [comment]
Though the easiest and most straightforward approach may be to just ask you to merge in master every now and then, and if there's a conflict we can submit it ourselves as a PR then rebase out other PRs on top of that. 

Food for thought on how you might do this type of stuff going forward, in any case a merge of master into this branch now would be helpful.

**Comment by dhermes** [comment]
@waprin I can just

```
git co logging-stdlib-handler-feature
git rebase master
```

and deal with conflicts. Have you tried it? Are there issues?

**Comment by dhermes** [comment]
@waprin I just got your branch up to date for you.

**Comment by waprin** [comment]
Thanks. I didn't try in this case and would have to wait till we get out of date to try again, but I did it in a previous PR. If I rebase on master, then my PR will be based on master so will have all those commits that origin/std-logging-feature doesn't, so when I make the PR all those master commits are included as part of the PR which we don't want. We really want to rebase the entire feature branch on master, or merge master into the feature branch, so when the PR is made you only see the actually new commits. But in order to rebase origin/std-logging-feature onto master, we need someone with write access to origin. Hopefully that makes sense.

**Comment by dhermes** [comment]
@waprin Just file an issue if you need your branch gardened

**Pull request closed without merging**